File: src/main/java/de/thkoeln/inf/gpm/vgb/model/Location.java
package de.thkoeln.inf.gpm.vgb.model;

import model.DiseaseDao;
import model.LocationDao;
import util.DbUtil;

import java.util.List;

public class Location {
    private Integer id = null;
    private String zipCode;
    private String name;

    public static Location createOrUpdate(Location location) {
        return DbUtil.INSTANCE.runInTransaction(() -> LocationDao.Companion.save(location));
    }

    public static List<Location> findAll() {
        return DbUtil.INSTANCE.runInTransaction(LocationDao.Companion::findAll);
    }

    public static Location findById(Integer locationId) {
        return DbUtil.INSTANCE.runInTransaction(() -> LocationDao.Companion.get(locationId)).toLocation();
    }

    public static void delete(Integer locationId) {
        DbUtil.INSTANCE.runInTransaction(() -> LocationDao.Companion.delete(locationId));
    }

    public Location(Integer id, String zipCode, String name) {
        this.id = id;
        this.zipCode = zipCode;
        this.name = name;
    }

    public Location(String zipCode, String name) {
        this.zipCode = zipCode;
        this.name = name;
    }

    public Integer getId() {
        return id;
    }

    public String getZipCode() {
        return zipCode;
    }

    public void setZipCode(String zipCode) {
        this.zipCode = zipCode;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}

File: src/main/java/de/thkoeln/inf/gpm/vgb/model/Location.kt
package de.thkoeln.inf.gpm.vgb.model

import org.jetbrains.exposed.dao.IntEntity
import org.jetbrains.exposed.dao.IntEntityClass
import org.jetbrains.exposed.dao.id.EntityID
import org.jetbrains.exposed.dao.id.IntIdTable
import org.jetbrains.exposed.sql.StdOutSqlLogger
import org.jetbrains.exposed.sql.addLogger
import org.jetbrains.exposed.sql.transactions.transaction

class Location(val id: Int? = null, val zipCode: String, val name: String)

class LocationDao(id: EntityID<Int>) : IntEntity(id) {
    private var zipCode by Locations.zipCode
    private var name by Locations.name

    companion object : IntEntityClass<LocationDao>(Locations) {
        /**
         * Update or create [Location] in database
         */
        fun save(location: Location) : Location? = transaction {
            val newLocation = if (location.id == null) {
                new { update(location) }
            } else {
                val old = LocationDao.findById(location.id)
                old?.update(location)
                LocationDao.findById(location.id)
            }

            newLocation?.toLocation()
        }
    }

    private fun update(location: Location) {
        this.zipCode = location.zipCode
        this.name = location.name
    }

    fun toLocation() = Location(id.value, zipCode, name)
}

object Locations : IntIdTable() {
    val zipCode = varchar("plz", 5)
    val name = varchar("name", 255)
}